### ISSUE TITLE
Parsing CSV with units in the header

### ISSUE THREAD

**Issue by micrenda** [hgrecco/pint-pandas#166]
This is not really a bug report, but more a suggestion for who may have the same issue as me (however, I would not be angry if a similar solution would be integrated in pint-pandas).

Let suppose you have a CSV like this:
```
molecule,reduced_field [1e-21 * V * m^2],magboltz_drift_velocity [m / s],magboltz_drift_velocity_precision,bolsig_reduced_mobility [1 / V / m / s],bolsig_drift_velocity [m / s],bolsig_delta,bolsig_validation,betaboltz_drift_velocity [m / s],betaboltz_drift_velocity_stdev [m / s],betaboltz_delta,betaboltz_validation
Ar,0.1,1702.9999999999998,0.45999999999999996,1.647e+25,1647.0,-3.2883147386964144,OK,1806.69,623.853,6.0886670581327165,OK
Ar,0.12589254117941673,1811.9999999999998,0.42,1.394e+25,1755.046,-3.143156732891819,OK,1891.57,504.651,4.3912803532008935,OK
Ar,0.15848931924611134,1932.9999999999998,0.62,1.179e+25,1868.715,-3.32565959648215,OK,2209.5899999999997,619.583,14.30884635281946,NOK
Ar,0.19952623149688797,2047.9999999999998,0.43,9.957e+24,1986.4215000000002,-3.006762695312487,OK,2263.21,573.505,10.50830078125002,NO
```
And you want to read these data using Pandas, you can use this code:
```python
import pandas as pd
import pint
import pint_pandas
import re

def fix_pint_pandas_units(df):
    p = re.compile(r'^(.*)\s*\[(.*)\]\s*$')
    for column in df.columns:
        m = p.match(column)
        if m:
            name = m.group(1).strip()
            unit = m.group(2).strip()
            df.rename(columns={column: name}, inplace=True)
            df[name] = pd.Series(df[name], dtype='pint[' + unit + ']')
            
if __name__ == '__main__':
    df = pd.read_csv('results.csv')
    fix_pint_pandas_units(df)
    print(df.dtypes)
```

May be this could be integrated directly in pint-pandas, eventually activated by a flag.

**Comment by gdmcbain** [comment]
Indeed, I hit this problem too because the CSV were to be stored either in 

- [Confluence databases](<https://www.atlassian.com/software/confluence/databases>) or 
- [Microsoft Lists](<https://www.microsoft.com/en-us/microsoft-365/microsoft-lists>)

both of which insist on single-line headers.   Note that I am far from endorsing either of these products (don't get me started&hellip;), but they're out there and some of us have to work with them.

I had handrolled a solution much like the OP's; mine used a solidus as that's how I think of units, as in `reduced_field / (1e-21 * V * m **2)`, but that's a matter of taste, and I'd be content with the original.

I'd be up for proposing a PR and seeing it through to merging, if there's a taste for an alternative to multiline headers.  Apart from this, pint-pandas is awesome&mdash;thank you very much!

**Comment by gdmcbain** [comment]
By the way, when I run the code in the OP, I get

> `TypeError`: data type 'pint[1e-21 * V * m^2]' not understood

If I switch `pd.Series` to `pint_pandas.PintArray`, I get instead 

> `ValueError`: Unit expression cannot have a scaling factor.

which looks like [a pint thing](<https://github.com/hgrecco/pint/blob/dc94eab312a4a33917166a0d999bd50c81922824/pint/facets/plain/registry.py#L1256>).

Did scaling factors used to work?

**Comment by andrewgsavage** [comment]
> I'd be up for proposing a PR and seeing it through to merging, if there's a taste for an alternative to multiline headers. Apart from this, pint-pandas is awesome—thank you very much!

go for it. Although I think it's best to suggest a solution first - I feel like you can do the same in 2-3 lines of pandas by splitting the column headers into a multiindex, at which point I'm not sure it's worth including as a function (but is worth adding to docs)


> Did scaling factors used to work?

idk, I never tested it. looks like that error message was at least 3 years old

**Comment by gdmcbain** [comment]
That's right, only a few lines are required. For my use-case, which is to encourage a standard way of storing and imbibing a table of data like material properties as function of temperature (quite similar to the OP), those 2–3 lines would be identical each time and so I'm thinking DRY and do want them in a single function rather than having to copy'n'paste. The question is whether that function lives in pint-pandas or some thin layer that sits on top that I can import; I'm happy to support the former but can live with the latter.
   What I've knocked up so far, which works for the first three tables I've tried is 

https://github.com/gdmcbain/csvu/blob/3fe99eb29a17bfec873bbc1890e6759c12d7b028/src/csvu/__init__.py#L7-L18

```py
def read_csv(filepath: Path) -> pandas.DataFrame:
    raw = pandas.read_csv(filepath)
    units = {}
    for k, c in enumerate(raw):
        if "/" in c:
            column, unit = c.split("/", 1)
        else:
            column = c
            unit = "dimensionless"
        units[column.rstrip()] = f"pint[{unit.strip(' ()')}]"
    raw.columns = list(units.keys())
    return raw.astype(units)
```

Ha, obviously a work in progress; I see `k` isn't used so the enumeration can be elided. Also I expect shortcomings of generality to be exposed as I throw more and more tables at it.

**Comment by andrewgsavage** [comment]
my thinking is that this part here:
```python3
    for k, c in enumerate(raw):
        if "/" in c:
            column, unit = c.split("/", 1)
        else:
            column = c
            unit = "dimensionless"
        units[column.rstrip()] = f"pint[{unit.strip(' ()')}]"
```
is going to change for each user, as they may use a different seperator or brackets or whatever. so that should be a function for the user to supply, ie

```python3
def parse_raw_name(raw_name):
   ...
   return column, unit
```
or in your case

```python3
def parse_raw_name(c):
        if "/" in c:
            column, unit = c.split("/", 1)
        else:
            column = c
            unit = "dimensionless"
        column = column.rstrip()
        unit = unit.strip(' ()')
   return column, unit
```
once you have this function, you can do something like
```python3
df = pd.read_csv(filename)
df.columns = pd.MultiIndex.from_tuples([parse_raw_name(c) for c in df.columns])
df = df.quantify()
```

The benefit of doing it this way is it's a little easier to debug - errors will occur in `[parse_raw_name(c) for c in df.columns]` which users can see, rather than in a pint-pandas file somewhere

**Comment by andrewgsavage** [comment]
> For my use-case, which is to encourage a standard way of storing and imbibing a table of data like material properties as function of temperature

I've been doing the same thing so am sympathetic to the cause! 

I am thinking an arguement could be added to quantify for `parse_raw_name`. We could also make this a module variable `QUANTIFY_PARSE_RAW_NAME` so quantify defaults to parsing single line column headers using this function.

And presumably the same for writing.

**Comment by gdmcbain** [comment]
> is going to change for each user

Yes, which reminds me of Hadley Wickham's ["Tidy datasets are all alike, but every messy dataset is messy in its own way.”](https://r4ds.hadley.nz/data-tidy.html)

> . so that should be a function for the user to supply,

Yes, for sure.   ...Though I'm wondering about the prospects for a nice default, rendering easy things easy, after having made hard things possible.  At the moment, `pandas.read_csv` does a pretty good job of this for CSV -> DataFrame; it's got a lot options, but reasonable defaults and moreover once I know what those defaults are, if I have a say in what the CSV looks like, then I can write the CSV in the first place so that it'll read into pandas without needing explicit options.  Of course I don't presume to dictate what that CSV format should be that is read in with units by default (unless it's just in my module, rather than here).

> I've been doing the same thing so am sympathetic to the cause!

What does your CSV-with-units format look like?  I should provide some examples too, shouldn't I.  I'll scrub some up.

Thanks very much for your feedback and suggestions.  I'll incorporate them immediately into https://github.com/gdmcbain/csvu/issues/3 and report back here with progress.

**Comment by andrewgsavage** [comment]
the format I have is `variable name (unit)`
eg
```
plastic strain (strain), temperature (degC)
```

you'll also need to consider columns with no unit, eg 'tester name', which can be denoted by setting the unit to 'No Unit'

**Comment by gdmcbain** [comment]
Ta.

Strain's an interesting case; it's dimensionless but strains can be added, or a stress can be divided by a strain to give say a Young modulus or two strains divided to give a Poisson ratio, but one wouldn't add a strain and a mass fraction.  Does Pint distinguish such incommensurate dimensionless quantities?  That sounds almost type-theoretic, like Haskell, where in the [`units`](<https://github.com/goldfirere/units/blob/455396ebc7dc42d483263acfc03cf1580f646ac5/units-defs/Data/Dimensions/SI.hs#L29-L33>) they do things like:

```hs
-- | A plane angle is dimensionless; alternatively, it has dimension
-- length/length. It would be wrong to divide 2 meters by 1
-- meter and conclude that the quantity is 2 radians or degrees. To make
-- plane angle safe to use, we define it as a fundamental dimension.
declareDimension "PlaneAngle"
``` 

which is appealing, though I wasn't thinking of going that far here.

But I don't see strain in the [Default Pint units definition file](<https://github.com/hgrecco/pint/blob/master/pint/default_en.txt>), and when I throw it in Pint raises

> `pint.errors.UndefinedUnitError`: 'strain' is not defined in the unit registry

Do you have extra code to handle "plastic strain (strain)"?  Or is an unknown unit to be treated as a special way?

**Comment by gdmcbain** [comment]
Another data-point on idioms: ["Great Tables for Scientific Publishing"](<https://posit-dev.github.io/great-tables/blog/tables-for-scientific-publishing/>) separates the name and unit with a comma; e.g., 

> Reaction Rate Constant (298 K), cm&sup3; molecules<sup>–1</sup> s<sup>–1</sup> 

or

> Temperature, °C

Not ideal for CSV, but I suppose it might work if quoted. 

I'm also sticking to ASCII myself because various operating systems which don't need to be named still muck up Unicode.

**Comment by gdmcbain** [comment]
The [American Physical Society](<https://cdn.journals.aps.org/files/styleguide-pr.pdf>) recommends parentheses, like @andrewgsavage *supra*.  Looks like a good choice for the default then.

**Comment by gdmcbain** [comment]
Ah, yes, I see from the _IUPAC Green Book_ cited in hgrecco/pint#1704 the idea of [quantity calculus](<https://en.wikipedia.org/wiki/Quantity_calculus>) is what I was reaching for [above](<https://github.com/hgrecco/pint-pandas/issues/166#issuecomment-2741945829>) in writing `f"{name} / {unit}".  Ta.  So now we've got APS v. IUPAC; so many standards to choose from&hellip;.

**Comment by andrewgsavage** [comment]
I think it's reasonable to assume a format of `f"{name}{separator}{unit}{suffix}"`
where in your example, `separator = " / ("` and `suffix = )`

we could loop through the common separators and suffixes and use the first one that works. Also store which is the first that works, and use that for writing later.

what are your thoughts on this approach?

**Comment by gdmcbain** [comment]
Ah I can imagine so many things that can go wrong. The IUPAC solidus, besides being used for division of a quantity by its unit in the quantity calculus, might also mean disjunction, like an alternative, like a type-theoretic union, as in "oxide/nitride", or even a subdirectory, like "data/residue"; the APS parentheses might be explanatory, like "oxide (dry)". But yes, this is good, I think this will be best, and the user can be trusted to beware such troublesome cases.  I'll put this in tomorrow, after I check on the formatters sent to matplotlib in that other issue, reusing them if possible. Ta.

**Comment by andrewgsavage** [comment]
some documentation here
https://pint-pandas.readthedocs.io/en/latest/user/reading.html#single-row-headers
